### ISSUE TITLE
ValidationError doesn't preserve JSON field's types

### ISSUE THREAD

**Issue by Genarito** [encode/django-rest-framework#7532]
## Checklist

- [X] I have verified that that issue exists against the `master` branch of Django REST framework.
- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [X] This is not a usage question. (Those should be directed to the [discussion group](https://groups.google.com/forum/#!forum/django-rest-framework) instead.)
- [X] This cannot be dealt with as a third party library. (We prefer new functionality to be [in the form of third party libraries](https://www.django-rest-framework.org/community/third-party-packages/#about-third-party-packages) where possible.)
- [X] I have reduced the issue to the simplest possible case.
- [ ] I have included a failing test as a pull request. (If you are unable to do so we can still accept the issue.)

## Steps to reproduce

I have a Response object with some Enums which I use in the frontend to identify some specific issues. I'm returning that response raising a ValidationError:

```python
from enum import Enum

class ResponseCode(Enum):
    SUCCESS = 1
    ERROR = 2

class MyErrorCode(Enum):
    INVALID_FORMAT_NON_NUMERIC = 1
    OTHER = 2


class ResponseStatus:
    def __init__(self, code: Enum, message: str = '', internal_code: Enum = None):
        self.code = code.value
        self.message = message
        self.internal_code = internal_code.value if internal_code is not None else None

    def to_json(self):
        return self.__dict__

class MySerializer(serializers.ModelSerializer):
    class Meta:
        model = MyModel
        fields = '__all__'

    @staticmethod
    def validate_name(value: str):
        raise serializers.ValidationError({
	    'status': ResponseStatus(
		ResponseCode.ERROR,
			message='Invalid string',
			internal_code=MyErrorCode.INVALID_FORMAT_NON_NUMERIC
	    ).to_json(),
	})
```

The problem is that, in this case, the fields `code` and `internal_code` are integer, but when they are returned they're implicitly casted to string for some reason:

![Current result](https://i.imgur.com/dBUpbQG.png)

I have some views which has the same response but they're returned using [JsonResponse](https://docs.djangoproject.com/en/3.1/ref/request-response/#jsonresponse-objects) in `views.py`:

`JsonResponse({...}, safe=False)`

The result:

![Expected](https://i.imgur.com/rR1se4p.png)

And this works correctly, so I think it's a problem with the serialization applied in the ValidationError returning.

## Expected behavior

Preserve the primitives types of information in the returned JSON object as shown in the above img (with JsonResponse)

**Comment by wjiuhe** [comment]
> https://github.com/encode/django-rest-framework/blob/bb133522efaf6ce3ae8fdf1dec6cd79566cfd166/rest_framework/exceptions.py#L39

> https://github.com/django/django/blob/48b4bae983a1f9a73624fe62ef5bd1130b0dba39/django/utils/encoding.py#L48

I think the related code is here. Should or not handle the integer case before calling `force_str`?

**Comment by j-boivie** [comment]
I have a similar issue when raising an exception from a custom permissions class, where bools are not serialised to JSON correctly. I.e. this:
```python
[...]
raise PermissionDenied({"Error": True, "Access": False})
```
Results in
```json
{
    "Error": "True", 
    "Access": "False"
}
```
So I assume that the exception needs to handle bools as well before coercing into strings as @wjiuhe pointed out

**Comment by Genarito** [comment]
Is there any repo mantainer here to make a PR? In the case is not, could I make a PR in my free time, or is expected to be forced to string always?

**Comment by ChrisCrossCrash** [comment]
I'm also having this issue (wanting to use non-string primitive type in a response produced by `serializers.ValidationError`).

I have an input on my site that takes a list of license numbers. I'd like the response to look like this(with an error only on the third item in the list):
```json
"license_numbers": [
    null,
    null,
    "The license number must consist of two letters followed by six numbers."
]
```

But instead it looks like this:
```json
"license_numbers": [
    "None",
    "None",
    "The license number must consist of two letters followed by six numbers."
]
```

However, I'd like to point out [this line in the official docs](https://www.django-rest-framework.org/api-guide/exceptions/#inspecting-api-exceptions):
> `.detail` - Return the textual description of the error.

Emphasis on the word "textual".

Also, this behavior is baked into so many existing applications already. Changing it would break a lot of stuff.

However, it would be nice to have a special parameter like `json_parse_msg`:
```
raise serializers.ValidationError([None, None, "invalid"])
# responds with:
"license_numbers": [
    "None",
    "None",
    "invalid"
]

raise serializers.ValidationError(errors, json_parse_msg=True)
# responds with:
"license_numbers": [
    null,
    null,
    "invalid"
]
```

**Comment by auvipy** [comment]
Hello, can you please atleast contribute the failing test case for this issue?

**Comment by Genarito** [comment]
Yep, an optional parameter as @ChrisCrossCrash proposed would be nice!
